homeassistant/components/habitica/calendar.py

```diff
@@ -0,0 +1,261 @@
+"""Calendar platform for Habitica integration."""
+
+from __future__ import annotations
+
+from datetime import datetime, time, timedelta
+from enum import StrEnum
+
+from homeassistant.components.calendar import (
+    CalendarEntity,
+    CalendarEntityDescription,
+    CalendarEvent,
+)
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+from homeassistant.util import dt as dt_util
+
+from . import HabiticaConfigEntry
+from .coordinator import HabiticaDataUpdateCoordinator
+from .entity import HabiticaBase
+from .types import HabiticaTaskType
+from .util import build_rrule, to_date
+
+
+class HabiticaCalendar(StrEnum):
+    """Habitica calendars."""
+
+    DAILIES = "dailys"
+    TODOS = "todos"
+    REMINDERS = "reminders"
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    config_entry: HabiticaConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up the calendar platform."""
+    coordinator = config_entry.runtime_data
+
+    async_add_entities(
+        [
+            HabiticaTodosCalendarEntity(coordinator),
+            HabiticaDailiesCalendarEntity(coordinator),
+            HabiticaRemindersCalendarEntity(coordinator),
+        ]
+    )
+
+
+class HabiticaCalendarEntity(HabiticaBase, CalendarEntity):
+    """Base Habitica calendar entity."""
+
+    def __init__(
+        self,
+        coordinator: HabiticaDataUpdateCoordinator,
+    ) -> None:
+        """Initialize calendar entity."""
+        super().__init__(coordinator, self.entity_description)
+
+
+class HabiticaTodosCalendarEntity(HabiticaCalendarEntity):
+    """Habitica todos calendar entity."""
+
+    entity_description = CalendarEntityDescription(
+        key=HabiticaCalendar.TODOS,
+        translation_key=HabiticaCalendar.TODOS,
+    )
+
+    @property
+    def event(self) -> CalendarEvent | None:
+        """Return the next upcoming event."""
+
+        events = [
+            CalendarEvent(
+                start=start,
+                end=start,
```

Comment:
The end date should be exclusive (https://developers.home-assistant.io/docs/core/entity/calendar/#get-events) and must be after start, so +1 day here, and below too.

Reply:
CalendarEvent already handles this and adds 1 day so i thought it might be unnecessary

Reply:
It's there just to fix input bugs, and is shared with the input handling code for creating events from the service.